opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConcurrentConfigurationControllerTest.java

```diff
@@ -0,0 +1,188 @@
+        env.setDataRoot(repository.getDataRoot());
+
+
+        List<RepositoryInfo> repositoryInfos = new ArrayList<>();
+        Map<String, Project> projects = new TreeMap<>();
+
+        /*
+         * Prepare PROJECTS_COUNT git repositories, named project-{i} in the test repositories directory.
+         */
+        for (int i = 0; i < PROJECTS_COUNT; i++) {
+            Project project = new Project();
+            project.setName("project-" + i);
+            project.setPath("/project-" + i);
+            RepositoryInfo repo = new RepositoryInfo();
+            repo.setDirectoryNameRelative("/project-" + i);
+
+            projects.put("project-" + i, project);
+            repositoryInfos.add(repo);
+
+            // create the repository
+            FileUtils.copyDirectory(
+                    Paths.get(repository.getSourceRoot(), "git").toFile(),
+                    Paths.get(repository.getSourceRoot(), "project-" + i).toFile()
+            );
+        }
+
+        env.setRepositories(repositoryInfos);
+        env.setProjects(projects);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        repository.destroy();
+
+        env.setProjects(origProjects);
+        env.setRepositories(origRepositories);
+        env.setSourceRoot(origSourceRootPath);
+        env.setDataRoot(origDataRootPath);
+    }
+
+    @Test
+    @ConditionalRun(RepositoryInstalled.GitInstalled.class)
+    public void testConcurrentConfigurationReloads() throws InterruptedException, ExecutionException {
+        final ExecutorService threadPool = Executors.newFixedThreadPool(THREAD_COUNT);
+        List<Future<?>> futures = new LinkedList<>();
+
+        /*
+         * Now run setting a value in parallel which triggers configuration reload.
+         */
+        for (int i = 0; i < TASK_COUNT; i++) {
+            futures.add(threadPool.submit(() -> {
+                Response put = target("configuration")
+                        .path("projectsEnabled")
+                        .request()
+                        .put(Entity.text("true"));
+
+                Assert.assertEquals(204, put.getStatus());
+
+                assertTestedProjects();
+            }));
+        }
+
+        threadPool.shutdown();
+        threadPool.awaitTermination(100, TimeUnit.MINUTES);
+
+        for (Future<?> future : futures) {
+            future.get();
+        }
+    }
+
+    @Test
+    @ConditionalRun(RepositoryInstalled.GitInstalled.class)
+    public void testConcurrentCInvalidateRepositories() throws InterruptedException, ExecutionException {
+        final ExecutorService threadPool = Executors.newFixedThreadPool(THREAD_COUNT);
+        final List<RepositoryInfo> repositoryInfos = env.getRepositories();
+        List<Future<?>> futures = new LinkedList<>();
+
+        /*
+         * Now run apply config in parallel
+         */
+        for (int i = 0; i < TASK_COUNT; i++) {
+            futures.add(threadPool.submit(() -> {
```

Comment:
why do you actually need the futures ?

Reply:
because when you trigger `get` it will rethrow all exceptions from the thread (i. e. the asserts)

Reply:
I see. Can you add a comment there ?

Reply:
yes